### PR TITLE
Check goroutines in places related to recent found leaks

### PR DESCRIPTION
**Pull request by jsoriano**
Follows the strategy used in elastic/beats#10850 to check for goroutine leaks
on tests in some places related to the leaks investigated as part
of #9302.

Several things here:
* New helper to reuse the goroutine checker, it also prints now the
  goroutines dump on failure.
* Add goroutine checks for autodiscover tests
* Add goroutine checks for `CloseOnSignal` and `SubOutlet`
  tests (they detect the issues solved by #11263)
* Add goroutine checks for log input  tests (they detect issues solved
  by elastic/beats#12125 and #12164)

### CODE REVIEW

File: filebeat/input/log/input.go [hunk 1]
```diff
@@ -87,11 +88,21 @@ func NewInput(
...
 
 	// stateOut will only be unblocked if the beat is shut down.
 	// otherwise it can block on a full publisher pipeline, so state updates
 	// can be forwarded correctly to the registrar.
 	stateOut := channel.CloseOnSignal(channel.SubOutlet(out), context.BeatDone)
+	defer func() {
+		if cleanupNeeded {
+			stateOut.Close()
+		}
+	}()
```
Comment (jsoriano) [inline filebeat/input/log/input.go]: These outlets should be probably created on `Run()`, but they are needed now to check the configuration. I'd fix it like this by now, and would do further refactors on future PRs.

File: libbeat/tests/resources/goroutines.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	return GoroutinesChecker{
+		before: runtime.NumGoroutine(),
+	}
+}
+
+// Check if the number of goroutines has increased since the checker
+// was created
+func (c GoroutinesChecker) Check(t testing.TB) {
+	t.Helper()
+	runtime.GC()
```
Comment (andrewkroh) [inline libbeat/tests/resources/goroutines.go]: What's the reasoning for triggering a GC?

Comment (jsoriano) [reply libbeat/tests/resources/goroutines.go]: I saw some goroutines were being counted after they seemed to be finished, adding sleeps on the checks didn't helped, but GC did, so I thought that GC was needed.
But I have tried again to give a time to goroutines to finish and it works without GC.

**Comment by jsoriano** [comment]
I am moving the fix for the leak in `NewInput` to its own PR (https://github.com/elastic/beats/pull/12125), so we can continue discussing about these tests here without blocking this fix.

**Comment by jsoriano** [comment]
> Do our runtime metrics that get logged every 30s include a gauge for the number of goroutines? That would be helpful for identifying if a leak is introduced to a Beat.

Created PR for that https://github.com/elastic/beats/pull/12135

File: filebeat/input/log/input.go [hunk 3]
```diff
@@ -727,14 +731,29 @@ func (p *Input) Wait() {
...
+
+		// stop all communication between harvesters and publisher pipeline
+		p.outlet.Close()
+	})
+}
 
-	// close state updater
-	p.stateOutlet.Close()
+// stopWhenDone takes care of stopping the input if some of the contexts are done
+func (p *Input) stopWhenDone() {
```
Comment (jsoriano) [inline filebeat/input/log/input.go]: PR for this in https://github.com/elastic/beats/pull/12164

**Comment by jsoriano** [comment]
Failing tests caused by https://github.com/elastic/beats/pull/12171

**Comment by jsoriano** [comment]
@exekias @andrewkroh this would be ready for review, thanks!

**Comment by jsoriano** [comment]
Test failure in Windows looks related, I'll look at this

File: libbeat/tests/resources/goroutines.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	var after int
+	for time.Now().Before(timeout) {
+		after = runtime.NumGoroutine()
+		if after <= c.before {
+			return nil
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+	profile := pprof.Lookup("goroutine")
+	profile.WriteTo(os.Stdout, 2)
```
Comment (andrewkroh) [inline libbeat/tests/resources/goroutines.go]: Nice. That's helpful.

**Pull request merged**